pkg/apiserver/rest/dualwriter_mode2.go

```diff
@@ -90,33 +93,48 @@ func (d *DualWriterMode2) List(ctx context.Context, options *metainternalversion
 		return nil, err
 	}
 
-	sl, err := d.Storage.List(ctx, options)
+	originKeys := []string{}
+	indexMap := map[string]int{}
+	for i, obj := range legacyList {
+		metaAccessor, err := utils.MetaAccessor(obj)
+		if err != nil {
+			return nil, err
+		}
+		originKeys = append(originKeys, metaAccessor.GetOriginKey())
+
+		accessor, err := meta.Accessor(obj)
+		if err != nil {
+			return nil, err
+		}
+		// Record the index of each LegacyStorage object so it can later be replaced by
+		// an equivalent Storage object if it exists.
+		indexMap[accessor.GetName()] = i
+	}
+
+	r, err := labels.NewRequirement(utils.AnnoKeyOriginKey, selection.In, originKeys)
 	if err != nil {
 		return nil, err
 	}
-	storageList, err := meta.ExtractList(sl)
+	optionsStorage := metainternalversion.ListOptions{
+		LabelSelector: labels.NewSelector().Add(*r),
+	}
+
+	sl, err := d.Storage.List(ctx, &optionsStorage)
 	if err != nil {
 		return nil, err
 	}
-
-	m := map[string]int{}
-	for i, obj := range storageList {
-		accessor, err := meta.Accessor(obj)
-		if err != nil {
-			return nil, err
-		}
-		m[accessor.GetName()] = i
+	storageList, err := meta.ExtractList(sl)
+	if err != nil {
+		return nil, err
 	}
 
-	for i, obj := range legacyList {
+	for _, obj := range storageList {
 		accessor, err := meta.Accessor(obj)
 		if err != nil {
 			return nil, err
 		}
-		// Replace the LegacyStorage object if there's a corresponding entry in Storage.
-		if index, ok := m[accessor.GetName()]; ok {
-			legacyList[i] = storageList[index]
-		}
+		legacyIndex := indexMap[accessor.GetName()]
+		legacyList[legacyIndex] = obj
```

Comment:
We shouldn't run into the situation where the name doesn't exist in the indexMap, but it doesn't hurt to check for it vs just overwriting the first element if it doesn't exist for some reason.
```suggestion
		if legacyIndex, ok := indexMap[accessor.GetName()]; ok {
			legacyList[legacyIndex] = obj
		}
```